CONTRIBUTING.md

```diff
@@ -0,0 +1,115 @@
+For more extensive changes, it is encouraged that you create it if one does not already exist.
+If another user has the issue assigned, coordinate your work with them.
+
+## Code Style
+
+Coding tackles the thorny problem of trying to express logical, computational concepts in a
+human-readable and understandable manner. Our code style favours the human side.
+We seek to find the balance between sections of code that are too large or dense to
+comprehend and the problem of too many small pieces of code to understand at once.
+
+Formatting is derived from Python's [PEP8](https://peps.python.org/pep-0008/) as implemented by
+[Black](https://pypi.org/project/black/), with line lengths limited to 100 (instead of the traditional 80).
+
+The linting toolchain uses
+[Black](https://pypi.org/project/black/),
+[Flake8](https://flake8.pycqa.org/en/latest/),
+[mypy](http://www.mypy-lang.org/), and
+[pylint](https://pylint.pycqa.org/en/latest/)
+to promote best practices for expressing the logical constructs.
+
+Expressing intent is the other key ingredient.
+Name code elements (classes, functions, variables) based on what purpose they serve,
+and try to organise code to convey the flow of logic.
+If there is additional information, use single-line comments.
+
+The "doc-block" at the top of class and functions should start with a quick summary of the purpose and intent.
+This summary can be followed by any relevant commentary on the implementation,
+such as links to decisions about the structure of the code and notes features that might not be obvious.
+
+The use of comments to note problems or future fixes is discouraged, and merges containing them will not be approved.
+Instead, create an issue to track the required change.
+
+<!--
+## Testing
+This section is to be drafted by whoever solves the issue relating to testing.
+
+-->
+
+## Branches And Commits
+
+The primary branch of this project is called 'mainline'.
+It represents the total summation of approved code changes.
+The tip of this branch will pass all linting and test stages.
+Tested stable versions are marked with git tags; there is no 'stable' branch.
+
+The mainline branch uses a 'semi-linear' branch history, with merge commits only being
+used to indicate the start and end of each applied topic branch.
+Rebasing code in topic branches is encouraged.
+
+Each commit in the repo should represent a logical, testable change.
+Ideally, they should also pass all linting and testing checks.
+When a branch is in active development, this does not need to be strictly adhered to,
+and you may wish to hold off on deciding on the commit order until the merge request.
+
+Commit messages, like code elements, starts with a summary of the intent of the change,
+and the full version can also include references to the issue(s) it solves and descriptions
+of the functional details of the changes.
+
+## Pull Requests: Getting Stuff Into The Project
```

Comment:
Do we have enforced merge requirements to prevent errant pull requests being solved by non-maintainers/when all reviewers have not say "yay"

Reply:
Yes; code review requirements are currently:
 - one maintainer approval
 - conversations in the merge resolved (which doesn't mean much because anyone can resolve conversations, iirc)
 - linting and sonar passing